java/arcs/core/analysis/RecipeGraph.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.HandleMode
+import arcs.core.data.Recipe
+
+/**
+ * A graph capturing the connections between particles and handles in [recipe].
+ *
+ * Nodes in the graph are either a [Recipe.Handle] or [Recipe.Particle]. For every write connection
+ * from a particle `p` to a handle `h` using a connection spec `s`, there is a labeled edge
+ * `p -s-> h` in the graph. Similarly, for every read connection from a particle `p` to a handle `h`
+ * using a connection spec `s`, there is a labeled edge `h -s-> p` in the graph.
+*/
+class RecipeGraph(recipe: Recipe) {
+    var nodes = mutableListOf<Node>()
+
+    init {
+        val handleNodesMap: Map<String, Node> = recipe.handles
+            .map { (_, handle) -> Node.Handle(handle) }
+            .associateBy { it.handle.name }
+        nodes = handleNodesMap.values.toMutableList()
+        recipe.particles.forEach { it.addNodeAndEdges(handleNodesMap) }
+    }
+
+    /** Adds a [Node.Particle] node and the edges induced by the handle connections. */
+    private fun Recipe.Particle.addNodeAndEdges(handleNodesMap: Map<String, Node>) {
+        val particleNode = Node.Particle(this)
+        nodes.add(particleNode)
```

Comment:
this seems like a side-effect, rather than part of the intended purpose of the function... could this be lifted into the `forEach`?

Reply:
Made this return the `particleNode` and lifted it to the caller.